reactioncommerce/reaction
# feat: use component library's CatalogGrid

Resolves #4583 and #4639 
Impact: **minor**  
Type: **feature**

## Solution
This PR replaces the current customer grid UI with the new `CatalogGrid` component.

## Breaking changes
Only breaking for custom sites as the customer-facing `productGridItem` component has been removed.

## Testing
1. Use reaction-data-loader on its 1.14 branch to load some products: `npx babel-node data/generator.js -m mongodb://localhost:3001/ -l dev -n meteor --max_old_space_size=8192 --presets es2015,stage-2`
2. Confirm grid with new styles works as expected - progressively loads images, responsive from mobile up, loading more as you scroll to the bottom
3. Currency - Change the currency via the dropdown in the header. Confirm prices are converted and displayed correctly, including prices where only 1 currency symbol should be displayed with a price range (RUB)
4. Search - Confirm improved search modal that uses the new grid, displays images (even when logged out), is perfectly responsive at any size, shows correct currency, and brings you to the PDP on click. NOTE: If using the data loader, the correct primary image may not be displayed due to a bug: https://github.com/reactioncommerce/reaction-data-loader/issues/8

### Badge label testing:
Note: Might be easier to create a new product and copy the PDP URL. If you create a product and try to find it in the admin UI (after loading the dev dataset), it'll take a while since the customer grid is sorted by newest first, but the admin grid is not.

1. Create a product, add title to variant, set quantity to 1, warn at 1. Publish & logout
2. Confirm “Limited supply” badge
3. Switch to other languages and confirm badge text updates reactively
4. Login and edit product, set variant quantity to 0, turn allow backorder off. Publish & logout
5. Confirm “Sold out” badge. Confirm translations
6. Login and turn allow backorder on for product. Publish & logout
7. Confirm “Backorder” label & translations

### Price range & search record update testing
Two core bugs were discovered while work on this PR: `product.price` was not always being updated, and search records were being updated before product changes were published. To confirm the fixes:
1. Create a product, add a title to the variant, set quantity, don't publish
2. Confirm the product doesn't appear in search yet
3. Publish the product
4. Confirm the product now does appear in search + the price range is correct, both in the main customer grid and search results
5. Edit the price for the variant and/or add a new visible variant with a price
6. Confirm in DB product.price range, min, and max are correct, but that the changes haven't propagated to search results yet
7. Publish the changes and confirm the search result's price range has updated
8. Confirm 6 & 7 with variants that have options.
